app/src/main/java/org/p2p/wallet/home/ui/main/HomePresenterDeeplinkHandler.kt

```diff
@@ -0,0 +1,79 @@
+package org.p2p.wallet.home.ui.main
+
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.launch
+import org.p2p.wallet.deeplinks.DeeplinkData
+import org.p2p.wallet.deeplinks.DeeplinkTarget
+import org.p2p.wallet.home.ui.main.models.HomeScreenViewState
+import org.p2p.wallet.infrastructure.coroutines.waitForCondition
+import org.p2p.wallet.newsend.ui.SearchOpenedFromScreen
+import org.p2p.wallet.swap.ui.orca.SwapOpenedFrom
+import org.p2p.wallet.user.interactor.UserInteractor
+
+/**
+ * Handles deeplinks supported by the home screen
+ */
+internal class HomePresenterDeeplinkHandler(
+    private val coroutineScope: CoroutineScope,
+    private val presenter: HomeContract.Presenter,
+    private val state: () -> HomeScreenViewState,
+    private val view: () -> HomeContract.View?,
+    private val userInteractor: () -> UserInteractor
+) {
+
+    suspend fun handle(data: DeeplinkData) {
+        when (data.target) {
+            DeeplinkTarget.BUY -> handleBuyDeeplink(data)
+            DeeplinkTarget.SEND -> handleSendDeeplink(data)
+            DeeplinkTarget.SWAP -> handleSwapDeeplink(data)
+            DeeplinkTarget.CASH_OUT -> handleCashOutDeeplink()
+            else -> Unit
+        }
+    }
+
+    private fun handleCashOutDeeplink() {
+        view()?.showCashOut()
+    }
+
+    private fun handleSwapDeeplink(data: DeeplinkData) {
+        if (data.args.containsKey("from") && data.args.containsKey("to")) {
+            val amount = data.args["amount"]?.toBigDecimalOrNull()?.toPlainString() ?: "0"
+            view()?.showSwapWithArgs(
+                tokenASymbol = data.args.getValue("from"),
+                tokenBSymbol = data.args.getValue("to"),
+                amountA = amount,
+                source = SwapOpenedFrom.MAIN_SCREEN
+            )
+        } else {
+            view()?.showSwap(SwapOpenedFrom.MAIN_SCREEN)
+        }
+    }
+
+    private suspend fun handleSendDeeplink(data: DeeplinkData) {
+        // fixme hack! waiting for tokens to load, probably it's better to show progress
+        waitForCondition(1000) { state().tokens.isNotEmpty() }
+        presenter.onSendClicked(SearchOpenedFromScreen.MAIN)
+    }
+
+    /**
+     * Handles buy token for fiat deeplink
+     */
+    private fun handleBuyDeeplink(data: DeeplinkData) {
+        val cryptoToken = data.args["to"]
+        val fiatToken = data.args["from"]
+        val fiatAmount = data.args["amount"]
+
+        if (!cryptoToken.isNullOrBlank() && !fiatToken.isNullOrBlank()) {
+            coroutineScope.launch {
+                val token = userInteractor().getSingleTokenForBuy(listOf(cryptoToken))
```

Comment:
if you want, you can simplify this class removing all the dependencies except CoroutineScope and View passing all the needed data as params
it explicitly shows what data method needs without looking into impl

Reply:
@gslevinkov  Do you mean making the private handle* functions open and calling them from the `when` statement inside the presenter, while passing the necessary arguments?